src/io/read_tree_genesis_hdf5.c

```diff
@@ -0,0 +1,1176 @@
+                }                                                       \
+                local_halos[i].sage_name = (int32_t) macro_forest_local_index; \
+            }                                                           \
+    }
+
+    int64_t *forest_offsets = gen->halo_offset_per_snap;
+    hid_t h5_fd = gen->h5_fds[filenum];
+    for(int isnap=end_snap;isnap>=start_snap;isnap--) {
+        hsize_t snap_offset[1], nhalos_snap[1];
+        snap_offset[0] = (hsize_t) forest_offsets[isnap];
+        nhalos_snap[0] = nhalos_per_snap[isnap];
+        if(nhalos_snap[0] == 0) continue;
+
+        /* fprintf(stderr,"forestnr = %"PRId64" reading nhalos = %llu at or snapshot = %d\n", forestnr, nhalos_snap[0], isnap); */
+
+        char snap_group_name[MAX_STRING_LEN];
+        snprintf(snap_group_name, MAX_STRING_LEN-1, "Snap_%03d", isnap);
+        hid_t h5_grp = H5Gopen(h5_fd, snap_group_name, H5P_DEFAULT);
+        XRETURN(h5_grp >= 0, -HDF5_ERROR, "Error: Could not open group = `%s` corresponding to snapshot = %d\n",
+                snap_group_name, isnap);
+
+        /* Merger Tree Pointers */
+        //Descendant, FirstProgenitor, NextProgenitor, FirstHaloInFOFgroup, NextHaloInFOFgroup
+        READ_PARTIAL_1D_ARRAY(h5_grp, isnap, head_enum, snap_offset, nhalos_snap, buffer);
+        /* Can not directly assign since 'Head' contains Descendant haloid which is too large to be contained
+           within 32 bits. I will need a separate assignment to break up haloid into a local index + snapshot,
+           and then use the forest-local offset for each snapshot */
+
+        const int is_mergertree_idx = 1, is_hosthaloid = 1;
+        ASSIGN_BUFFER_WITH_MERGERTREE_IDX_TO_SAGE(nhalos_snap, int64_t, Descendant, isnap, is_mergertree_idx, ~is_hosthaloid);
+
+        //Same with 'Tail' -> 'FirstProgenitor'
+        READ_PARTIAL_1D_ARRAY(h5_grp, isnap, tail_enum, snap_offset, nhalos_snap, buffer);
+        ASSIGN_BUFFER_WITH_MERGERTREE_IDX_TO_SAGE(nhalos_snap, int64_t, FirstProgenitor, isnap, is_mergertree_idx, ~is_hosthaloid);
+
+        //And same with 'hostHaloID' -> FirstHaloinFOFGroup.
+        READ_PARTIAL_1D_ARRAY(h5_grp, isnap, hosthaloid_enum, snap_offset, nhalos_snap, buffer);
+        ASSIGN_BUFFER_WITH_MERGERTREE_IDX_TO_SAGE(nhalos_snap, int64_t, FirstHaloInFOFgroup, isnap, ~is_mergertree_idx, is_hosthaloid);
+
+
+        /* MS 3rd June, 2019: The LHaloTree convention (which sage uses) is that Mvir contains M200c. While this is DEEPLY confusing,
+           I am using C 'unions' to reduce the confusion slightly. What will happen here is that 'Mass_200crit' will get
+           assigned to the 'M200c' field within the halo struct; but that 'M200c' will also be accessible via 'Mvir'.
+        */
+        READ_PARTIAL_1D_ARRAY(h5_grp, isnap, m200c_enum, snap_offset, nhalos_snap, buffer);
+        ASSIGN_BUFFER_TO_SAGE(nhalos_snap, double, M200c, float);//M200c is an alias for Mvir
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Casting between integer* and double* which have an incompatible binary data representation.](https://app.codacy.com/manual/jacobseiler/sage-model/pullRequest?prid=4620074)